### ISSUE TITLE
Blaze

### ISSUE THREAD

**Issue by tmeasday** [meteor/guide#42]
### Article outline

https://github.com/meteor/guide/blob/master/outlines/blaze.md

### Major decision points

1. Blaze templates should be used as encapsulated components, using the following strategies:
  1. Validation of data context and always passing it explicitly on template inclusion
  2. Explicitly naming variables, contexts, and helpers
  3. Storing internal state on the template instance
2. Developers should make a clear distinction between "smart" components that load data and access state, and "pure" components that just render UI

--------

Previous proposed outline:

1. Introduction to spacebars, Blaze's templating syntax
 - Basic builtis
 - Template lifecycle
2. Making a Blaze template a reusable component:
  - Being explicit about data contexts
  - Never use `this.title`, instead `post.title` (does the new iterator syntax help with this?)
  - Storing state on the template instance, patterns for access in helpers/spacebars
  - Being careful about using `this.$` and event maps
  - Passing callbacks as arguments
  - How to use a component as a layout and `Template.contentBlock`
  - Integrating w/ 3rd party libraries (i.e. jquery)
3. Accessing data from a template
  - Template level subscriptions
  - Accessing collections
  - Accessing stores from templates
  - Understanding reactivity
  - Pure vs impure components
4. Reusing code between templates
  - Global helpers (and what they should be used for)
  - Utility code
6. Testing Blaze components (or is this in the testing chapter?)

**Comment by stubailo** [comment]
Original outline: 

### Blaze guide: The Tracker-based reactive templating system

Write “HTML with holes” just like you're used to, and get a fast, fine-grained, reactively updating page with no sweat.

1. Spacebars syntax, and how to use the built-in helpers
2. Building reusable components with Blaze by avoiding global data
3. How to use reactivity in a principled way
4. Writing maintainable helpers and event handlers that aren't tightly coupled to HTML
5. Reusing logic and HTML snippets between templates

**Comment by VladShcherbin** [comment]
Maybe a small chapter with a comparison with React/Angular as lots of users are asking this all the time.

**Comment by stubailo** [comment]
Perhaps. However, I don't think the guide is a place for discussion - that will clutter it up for people who are just looking for answers. We should have this content somewhere, and link to it from this guide.

**Comment by VladShcherbin** [comment]
@stubailo , sure, I also think that the Blaze guide is not the best part. Maybe we can have something like an overview of the front-end frameworks for Meteor, their pros and cons, differences, etc.

Don't know, where to put this though but it can be a great part of the guide as many users are seeking for this answer, both newcomers and experienced ones.

**Comment by stubailo** [comment]
Oooh, that's a good point - that might be a big decision for a lot of people when building a serious app. Framed in that way, it actually makes a lot of sense for that to be in the guide, much like a discussion of different databases would also belong if we supported more than Mongo.

**Comment by tmeasday** [comment]
Ok, posted a next iteration on what I think the headings are.

I don't think we are decided yet on whether we be more boilerplatey like @sanjo's comment here: https://github.com/meteor/guide/issues/8#issuecomment-146110276 or tell people to use Blaze Components.

Perhaps I will try both in the todos app and see how it feels.

(Keeping this separate from the other ticket which is really a discussion on whether to use Blaze Components. In fact I'll rename that ticket).

**Comment by stubailo** [comment]
> Accessing data from a template

This seems to be more in line with the data loading chapter, if we are suggesting loading data pretty much only in templates?

I also think the beginning should go over how the thing works a little? Like somewhere there should be a description of the following:

1. What is the data context/what's a View
2. How does the build system translate `.html` into JavaScript using HTMLJS
3. What parts of the page re-render when data changes

Basically, some stuff that you would need to know to do advanced work with Blaze.

Also, what about animations/uihooks/momentum?

**Comment by stubailo** [comment]
https://github.com/meteor/meteor/issues/4977

It could be good to go over how array reactivity is supposed to work, as well.

**Comment by mitar** [comment]
> Perhaps I will try both in the todos app and see how it feels.

Yes, that would be great. :-) I would love your feedback.

**Comment by stubailo** [comment]
> Never use this.title, instead post.title

You mean like always use `{{#let}}` and `{{#each ... in}}`?

> Being careful about using this.$ and event maps

I think this should also strongly caution against using global `$` - people do this all the time

> Testing Blaze components (or is this in the testing chapter?)

I think this should be in the testing chapter

### Other things

1. Helpers should not have side effects, and should expect to rerun often, so don't put high-CPU code there
2. I really like the https://github.com/raix/Meteor-handlebar-helpers package - I think it is a wonderful workaround for Meteor's lack of JS expressions in templates, IMO it would be good to recommend it
3. We should have some great patterns for showing loading indicators, or at least a snippet of example code. `this.subscribe` makes it easy.

**Comment by tmeasday** [comment]
> > Never use this.title, instead post.title

> You mean like always use {{#let}} and {{#each ... in}}?

Yup.

> > Being careful about using this.$ and event maps

> I think this should also strongly caution against using global $ - people do this all the time

That's what I meant, more or less

> > Testing Blaze components (or is this in the testing chapter?)

> I think this should be in the testing chapter

I guess I was thinking about the mechanics of rendering a component to a off-screen div and using selectors to find HTML. It's Blaze-specific so wouldn't be in the testing chapter if we wanted to keep it agnostic?

Other things

> 1. Helpers should not have side effects, and should expect to rerun often, so don't put high-CPU code there

Good point, very much need to mention this. Perhaps a "reactivity" section could mention this and the each thing?

> I really like the https://github.com/raix/Meteor-handlebar-helpers package - I think it is a wonderful workaround for Meteor's lack of JS expressions in templates, IMO it would be good to recommend it

I'm not sure I'd recommend it to the point of using in the example apps but maybe a list of "useful packages for Blaze" section could prominently feature it?

> We should have some great patterns for showing loading indicators, or at least a snippet of example code. this.subscribe makes it easy.

Yeah. This is a section in the UI/UX chapter, but we could mention `Template.subscriptionsReady` here.

**Comment by stubailo** [comment]
> I'm not sure I'd recommend it to the point of using in the example apps

Not sure why? It basically fills in for a missing feature of having expressions in Spacebars.

**Comment by tmeasday** [comment]
It's just a bit messy is all

**Comment by mitar** [comment]
> It's just a bit messy is all

+ 1 :-)

It is cool inspiration, but I mostly just copy functions from there.

> Helpers should not have side effects, and should expect to rerun often, so don't put high-CPU code there

Maybe mentioning things like [computed field](https://github.com/peerlibrary/meteor-computed-field) to minimize/cache reactivity as well?

> I guess I was thinking about the mechanics of rendering a component to a off-screen div and using selectors to find HTML. It's Blaze-specific so wouldn't be in the testing chapter if we wanted to keep it agnostic?

Maybe tinytest could be extended with things like `trimEquals` (like [here](https://github.com/peerlibrary/meteor-blaze-components/blob/master/tests/tests.coffee#L1)) so that it is easier to compare HTML?

And also testing things for real components, so when you do [need to test events](https://github.com/peerlibrary/meteor-blaze-components/blob/master/tests/tests.coffee#L1028).

**Comment by Slava** [comment]
In the current deployed version of the guide I have noticed this text:

> XXX: we are violating our own rule here and reading into the sub-component. But there’s no real mechanism to pass through generic event handlers in Blaze

In the "Reusing code in Blaze - Composition" section. It is regarding this code:

```handlebars
<template name="autocompleteInput">
  {{> blurringInput name=name value=currentValue}}
</template>
```

```javascript
Template.autocompleteInput.helpers({
  currentValue() {
    // perform complex logic to determine the auto-complete's current text value
  }
});

Template.autocompleteInput.events({
  'change input': function(event, instance) {
    // read the current value out of the input, potentially change the value
  }
});
```

So a proposed "clean" solution is for every reusable component such as `blurringInput` to emit jQuery events when it is necessary and describe them in the components docs. So instead of listening on the `change input` event in the parent event-map, you would listen on `input-change componentEl` event or something like this. Granted, capturing the right DOM node might be hard (would be easier if Blaze forced attaching event-listeners on components).

**Comment by tmeasday** [comment]
@Slava - you're right! This is the "correct" way to do this in Blaze, I had forgotten. I guess a question is whether to do this, or the more "Reacty" thing of passing `onChange` in as an argument to `blurringInput`.

What do you think @stubailo?

**Comment by robfallows** [comment]
Just going through the content again and came across this: https://github.com/meteor/guide/blob/master/content/blaze.md#helpers-in-tags

Not having seen that style of helper before, I got all excited :wink:. But either I'm missing something fundamental, or it just doesn't work. I'm hoping for the former, because it looks really neat.

(I used your exact example)

**Comment by stubailo** [comment]
Hmm, it should definitely work. We should make sure that code sample is correct!

**Comment by reel** [comment]
It needs transpiling. @robfallows are you using the ecmascript package?

**Comment by robfallows** [comment]
@reel : yes - it's not the transpiling that was the issue - it built ok, but just didn't work.

**Comment by robfallows** [comment]
So, I've got a bit further with this. I get failure to render if I use (the example from the guide):

```javascript
Template.foo.helpers({
  attributes() {
    return {
      class: 'A class',
      style: {background: 'blue'}
    };
  }
});
```

However, it works fine without nested objects:

```javascript
Template.foo.helpers({
  attributes() {
    return {
      class: 'A class',
      style: 'background: blue;'
    };
  }
});
```

**Comment by robfallows** [comment]
Regarding this section: https://github.com/meteor/guide/blob/master/content/blaze.md#if--unless

> In JS `null`, `undefined`, `0`, `''`, `[]`, and `false` are considered "falsy", and all other values are "truthy".

We should maybe link somewhere that explains this more fully. We could. for example add `NaN` as a falsy value. However, my biggest problem is `[]` which evaluates to truthy or falsy depending on how you use it. `!![]` evaluates as `true`, whereas `[]==true` evaluates to `false`.

**Comment by stubailo** [comment]
Oh wow, interesting. `[]` might be a case where Blaze actually deviates from JS in what is truthy or falsy. That's unfortunate.
